pkg/porter/archive.go

```diff
@@ -170,26 +165,131 @@ func (ex *exporter) export() error {
-	err := filepath.Walk(path,
-		func(subpath string, info os.FileInfo, err error) error {
+func (ex *exporter) createTarHeader(ctx context.Context, path string, file string, fileInfo os.FileInfo) (*tar.Header, error) {
+	log := tracing.LoggerFromContext(ctx)
+
+	header := &tar.Header{
+		ModTime:    time.Unix(0, 0),
+		AccessTime: time.Unix(0, 0),
+		ChangeTime: time.Unix(0, 0),
+		Uid:        0,
+		Gid:        0,
+	}
+
+	switch {
+	case fileInfo.Mode().IsDir():
+		header.Typeflag = tar.TypeDir
+		header.Mode = 0755
+	case fileInfo.Mode().IsRegular():
+		header.Typeflag = tar.TypeReg
+		header.Mode = 0644
+		header.Size = fileInfo.Size()
+	default:
+		log.Debugf("Skipping %s. Not a file/dir", file)
+		return nil, nil
+	}
+
+	// ensure header has relative file path prepended with '.'
+	relativeFilePathName := file
+
+	if filepath.IsAbs(path) {
+		relativePath, err := filepath.Rel(path, file)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if relativePath != "." {
+			relativeFilePathName = fmt.Sprintf(".%s%s", string(filepath.Separator), relativePath)
+		} else {
+			relativeFilePathName = relativePath
+		}
+	}
+
+	log.Debugf("relativeFilePathName: %s\n", relativeFilePathName)
+
+	header.Name = filepath.ToSlash(relativeFilePathName)
+
+	// directories must be suffixed with '/'
+	if fileInfo.Mode().IsDir() && !strings.HasSuffix(header.Name, "/") {
+		header.Name += "/"
+	}
+
+	log.Debugf("header.Name: %s\n", header.Name)
+
+	return header, nil
+}
+
+func (ex *exporter) CustomTar(ctx context.Context, srcPath string) (io.ReadCloser, error) {
+	pipeReader, pipeWriter := io.Pipe()
+
+	gzipWriter := gzip.NewWriter(pipeWriter)
+	tarWriter := tar.NewWriter(gzipWriter)
+
+	path := filepath.Clean(filepath.Join(srcPath, "."))
+
+	go func() {
+		ctx, log := tracing.StartSpan(ctx)
+
+		defer func() {
+			if err := tarWriter.Close(); err != nil {
+				log.Warnf("Can't close tar writer: %s", err)
+			}
+			if err := gzipWriter.Close(); err != nil {
+				log.Warnf("Can't close gzip writer: %s\n", err)
+			}
+			if err := pipeWriter.Close(); err != nil {
+				log.Warnf("Can't close pipe writer: %s\n", err)
+			}
+			log.EndSpan()
+		}()
+
+		walker := func(file string, finfo os.FileInfo, err error) error {
 			if err != nil {
 				return err
 			}
-			err = ex.fs.Chtimes(subpath, time.Unix(0, 0), time.Unix(0, 0))
+
+			log.Debugf("Packaging directory %s\n", file)
+
+			hdr, err := ex.createTarHeader(ctx, path, file, finfo)
 			if err != nil {
 				return err
 			}
+
+			// if header is nil then it's not a regular file nor directory
+			if hdr == nil {
+				return nil
+			}
+
+			if err := tarWriter.WriteHeader(hdr); err != nil {
+				return err
+			}
+
+			// if path is a dir, nothing more to do
+			if finfo.Mode().IsDir() {
+				return nil
+			}
+
+			// add file to tar
+			sourceFile, err := os.Open(file)
+			if err != nil {
+				return err
+			}
+
+			defer sourceFile.Close()
+			_, err = io.Copy(tarWriter, sourceFile)
+			if err != nil {
+				return err
```

Comment:
In general it's helpful to wrap errors from external library calls to provide a bit more context of what we were doing when it failed. For example, here if io.Copy fails, adding some context about the file we were trying to write would make it easier to troubleshoot if it fails.

That way we get an error like "failed to write somefile.json: unexpected EOF" instead of just the EOF without knowing which one failed.

```suggestion
				return fmt.Errorf("failed to copy %s: %w", file, err)
```